### PR TITLE
Add "Autobuild", builder and filler roles, and room reboot

### PR DESCRIPTION
**Pull request by tedivm**
* Creates the concept of `practicalRoomLevel`, which can be compared against the real room level to see what structures should be built next. This means that a room being built up multiple levels won't just build up a specific building type, but will go from one room level to another until everything is complete. 

* The closest planned structures to the storage location are built first.

* Roads, Links, Walls, and Ramparts are not included in this.

* The `Builder` role is used to actually complete the structures. It will upgrade the controller if there isn't anything to build.

* A basic `Filler` role refills the spawns and extensions. Without this the rooms would die as they attempt to build creeps they can't actually support.

* A very basic "reboot" program has been added to spawn a 300 energy filler whenever the room is dead.

### CODE REVIEW

File: src/programs/city/reboot.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Provide Room-level Security
```
Comment (anisoptera) [inline src/programs/city/reboot.js]: Not really?

File: src/roles/filler.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+const MetaRole = require('roles_meta')
+
+class Builder extends MetaRole {
```
Comment (anisoptera) [inline src/roles/filler.js]: Not the right class name, I think :)

File: src/roles/filler.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      var target = spawns[0]
+    }
+    if (creep.pos.getRangeTo(target) > 3) {
+      creep.moveTo(target)
+    }
+
+  }
+}
+
+module.exports = Builder
```
Comment (anisoptera) [inline src/roles/filler.js]: Ditto here.

File: src/roles/meta.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+
+class MetaRole {
+
+  recharge(creep) {
```
Comment (anisoptera) [inline src/roles/meta.js]: Not sure why we need a base class for one function that seems pretty standalone. Also, I don't see the reason to have a base class at all, given the way roles work in this system. These are just helper functions, could be free functions in a module and callable from anywhere someone has a Creep.

Comment (tedivm) [reply src/roles/meta.js]: I definitely plan on expanding this out into other things though, particularly with regards to helper functions with spawning and boosting.

Comment (anisoptera) [reply src/roles/meta.js]: Right, but the thing I'm wondering about here is, if the class doesn't have any state, why do those methods need to go onto a base class, and not just a helper lib?

Sort of a stylistic nitpick, though.

**Pull request merged**